### PR TITLE
Tenant key fixes

### PR DESCRIPTION
**Pull request by pdeziel**
### Scope of changes

This attempts to fix the key issues based on the discussion in sc-14203.

Fixes SC-14203

### Type of change

- [ ] new feature
- [ ] bug fix
- [ ] documentation
- [ ] testing
- [x] technical debt
- [ ] other (describe)

### Acceptance criteria

Describe how reviewers can test this change to be sure that it works correctly. Add a checklist if possible.

### Author checklist

- [x] I have manually tested the change and/or added automation in the form of unit tests or integration tests
- [ ]  I have updated the dependencies list
- [ ]  I have recompiled and included new protocol buffers to reflect changes I made
- [ ]  I have added new test fixtures as needed to support added tests
- [ ]  Check this box if a reviewer can merge this pull request after approval (leave it unchecked if you want to do it yourself)
- [ ]  I have moved the associated Shortcut story to "Ready for Review"

### Reviewer(s) checklist

- [ ] Any new user-facing content that has been added for this PR has been QA'ed to ensure correct grammar, spelling, and understandability.
- [ ] Are there any TODOs in this PR that should be turned into stories?

### CODE REVIEW

File: pkg/tenant/organizations.go [hunk 1]
```diff
@@ -72,3 +73,25 @@ func (s *Server) OrganizationDetail(c *gin.Context) {
 	}
 	c.JSON(http.StatusOK, out)
 }
+
+// Helper to fetch the orgID from the gin context. This method also logs and returns
+// any errors to allow endpoints to have consistent error handling.
```
Comment (pdeziel) [inline pkg/tenant/organizations.go:78]: Do we have a follow on story to take care of the tech debt related to this?

File: pkg/tenant/db/topics_test.go [hunk 2]
```diff
@@ -90,8 +138,21 @@ func (s *dbTestSuite) TestCreateTopic() {
 	require.NoError(err, "could not validate topic data")
 
 	s.mock.OnPut = func(ctx context.Context, in *pb.PutRequest) (*pb.PutReply, error) {
-		if len(in.Key) == 0 || len(in.Value) == 0 || in.Namespace != db.TopicNamespace {
-			return nil, status.Error(codes.FailedPrecondition, "bad Put request")
+		switch len(in.Key) {
```
Comment (pdeziel) [inline pkg/tenant/db/topics_test.go:141]: Add a story to add a "multi-get" to the trtl mock to make testing with multiple gets easier.

File: pkg/tenant/api/v1/client.go [hunk 3]
```diff
@@ -518,10 +518,6 @@ func (s *APIv1) ProjectCreate(ctx context.Context, in *Project) (out *Project, e
 }
 
 func (s *APIv1) ProjectDetail(ctx context.Context, id string) (out *Project, err error) {
-	if id == "" {
-		return nil, ErrProjectIDRequired
-	}
```
Comment (bbengfort) [inline pkg/tenant/api/v1/client.go:523]: Unless I'm mistaken; If `id == ""` then the request will be made to `GET /v1/projects/` based on the `fmt.Sprintf` on line 521; the response will likely return a 200; but it won't contain a `*Project` but a `*ProjectPage` -- do we need to handle this error condition?

File: pkg/tenant/api/v1/client.go [hunk 4]
```diff
@@ -716,7 +704,7 @@ func (s *APIv1) ProjectAPIKeyList(ctx context.Context, id string, in *PageQuery)
 
 	// Make the HTTP request
 	var req *http.Request
-	if req, err = s.NewRequest(ctx, http.MethodGet, path, nil, &params); err != nil {
+	if req, err = s.NewRequest(ctx, http.MethodGet, path, in, &params); err != nil {
```
Comment (bbengfort) [inline pkg/tenant/api/v1/client.go:707]: This is a `GET` request so the request values are passed in via `&params` which is converted into query params on lines 700-703. Sending a body (which is the `nil`/`in` parameter) is not part of the `GET` semantic.

File: pkg/tenant/db/db.go [hunk 5]
```diff
@@ -162,6 +162,26 @@ type Model interface {
...
+	}
+
+	// Unmarshal the data into the model
+	if err = model.UnmarshalValue(value); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getRequest(ctx context.Context, namespace string, key []byte) (value []byte, err error) {
```
Comment (bbengfort) [inline pkg/tenant/db/db.go:184]: nicely broken up!

File: pkg/tenant/db/keys.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+// Helper to delete an object's key from the database.
+func DeleteObjectKey(ctx context.Context, key []byte) (err error) {
+	k := &Key{}
+	if err = k.UnmarshalValue(key); err != nil {
+		return err
+	}
+
+	return Delete(ctx, k)
+}
```
Comment (bbengfort) [inline pkg/tenant/db/keys.go:99]: Excellent module!

File: pkg/tenant/db/keys_test.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"testing"
+
+	"github.com/oklog/ulid/v2"
+	"github.com/rotationalio/ensign/pkg/tenant/db"
+	"github.com/stretchr/testify/require"
+)
+
+func TestKey(t *testing.T) {
```
Comment (bbengfort) [inline pkg/tenant/db/keys_test.go:11]:
```suggestion
func TestNullSanity(t *testing.T) {
    require.Len(t, db.NullID, 16, "unexpected null ID length") 
    require.Zero(t, db.NullID, "expected null ID to be zero valued") 
}

func TestKey(t *testing.T) {
```

File: pkg/tenant/organizations.go (hunk 1 above)
Comment (bbengfort) [reply pkg/tenant/organizations.go:78]: I was unable to find one.

**Review by bbengfort — APPROVED** [review]
A long slog, but well done on this PR @pdeziel!

File: pkg/tenant/api/v1/client.go (hunk 3 above)
Comment (pdeziel) [reply pkg/tenant/api/v1/client.go]: Thanks, this was leftover from a previous version of the PR.

File: pkg/tenant/api/v1/client.go (hunk 4 above)
Comment (pdeziel) [reply pkg/tenant/api/v1/client.go]: Thanks for catching this!

**Pull request merged**